### PR TITLE
Restructured Puddle & added cli input

### PR DESCRIPTION
**Pull request by AlexDunmow**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,5 +22,3 @@ _testmain.go
 *.exe
 *.test
 *.prof
-
-puddle
```
Comment (nii236) [inline .gitignore:26]: Why was this removed? There's a chance this might get checked into git

File: puddle/puddle.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		select {
+		case msg := <-rtm.IncomingEvents:
+			log.Print("Event Received: ")
+			switch ev := msg.Data.(type) {
+			case *slack.HelloEvent:
+				// Ignore hello
+
+			case *slack.ConnectedEvent:
+				log.Println("######### Connected to Slack #########")
+				// log.Println("Infos:", ev.Info)
```
Comment (nii236) [inline puddle/puddle.go:57]: Are these still needed?

File: puddle/puddle.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+				// log.Println("Connection counter:", ev.ConnectionCount)
+				// Replace #general with your Channel ID
+				rtm.SendMessage(rtm.NewOutgoingMessage(fmt.Sprintf("... and I'm back! Git tag: %s", GITTAG), CHANNEL))
+
+			case *slack.MessageEvent:
+				j, _ := json.Marshal(ev.Msg)
```
Comment (nii236) [inline puddle/puddle.go:63]: Should handle this error

File: puddle/puddle.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+				log.Printf("Error: %s\n", ev.Error())
+
+			case *slack.InvalidAuthEvent:
+				log.Printf("Invalid credentials")
+				break Loop
+
+			default:
+
+				// Ignore other events..
+				// log.Printf("Unexpected: %v\n", msg.Data)
```
Comment (nii236) [inline puddle/puddle.go:83]: Is this needed?

File: puddle/slack.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+)
+
+var lock = new(sync.Mutex)
+var adminCommands = make(map[string]func(string, string, slack.Msg) error)
+
+//Processes messages from Slack channel #puddle
```
Comment (nii236) [inline puddle/slack.go:18]: Should start with ProcessMessage...

File: puddle/slack.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+//Ideally, this should be a router
+func ProcessMessage(msg slack.Msg) error {
+	if strings.HasPrefix(msg.Text, "!") {
+		username, _ := GetUsername(msg)
+		ProcessAdminCommand(username, msg.Text, msg)
+	}
+	return nil
+}
+
+//Retreive username from SLACK API
```
Comment (nii236) [inline puddle/slack.go:28]: GetUsername...

**Comment by AlexDunmow** [comment]
Where's the comments?

File: puddle/slack.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	if err != nil {
+		fmt.Println(color.RedString("warning"), "No user information: ", msg.Text)
+		return "", err
+	}
+
+	return userInfo.Name, nil
+}
+
+//Parse message and search admin commands and feed message
+func ProcessAdminCommand(username string, msgText string, msg slack.Msg) {
```
Comment (nii236) [inline puddle/slack.go:43]: ProcessAdminCommand...

File: puddle/slack.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	cmdString := strings.TrimPrefix(words[0], "!")
+
+	if fn, ok := adminCommands[cmdString]; ok {
+		go fn(username, msgText, msg)
+	} else {
+		SendMessage(fmt.Sprintf("@%s Wat?", username))
+	}
+}
+
+//Simple send message to @username: <msg>
```
Comment (nii236) [inline puddle/slack.go:54]: SendMessage...

File: puddle/slack.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if ISDEV == "true" {
+		fmt.Println(">> ", text)
+		return nil
+	}
+	rtm.SendMessage(rtm.NewOutgoingMessage(text, CHANNEL))
+
+	return nil
+}
+
+//Pull latest git master and rebuild, restart puddlebot
```
Comment (nii236) [inline puddle/slack.go:65]: Build...

File: puddle/slack.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	SendMessage("...Restarting...")
+
+	cmd := exec.Command("./run.sh")
+	cmd.Start()
+
+	os.Exit(1)
+	return err
+}
+
+//Register admin command
```
Comment (nii236) [inline puddle/slack.go:89]: RegisterAdminCommand...

File: puddle/slack.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+	if _, ok := adminCommands[cmdString]; !ok {
+		adminCommands[cmdString] = fn
+		return nil
+	}
+
+	return errors.New("Command exists")
+}
+
+func init() {
+	RegisterAdminCommand("build", Build)
```
Comment (nii236) [inline puddle/slack.go:105]: How are admin commands different from normal commands?

File: puddle/slack.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+		fmt.Println(">> ", text)
+		return nil
+	}
+	rtm.SendMessage(rtm.NewOutgoingMessage(text, CHANNEL))
+
+	return nil
+}
+
+//Pull latest git master and rebuild, restart puddlebot
+func Build(username string, msgText string, msg slack.Msg) error {
```
Comment (nii236) [inline puddle/slack.go:66]: Should explicit user-facing commands each have their own package and be registered below?

**Review by jgunnink — CHANGES_REQUESTED** [review]
> Where's the comments?

Here you go:
1) No tests
2) CI is failing
3) No PR description.

😄

**Review by nii236 — CHANGES_REQUESTED** [review]

File: puddle/slack.go (hunk 11 above)
Comment (AlexDunmow) [reply puddle/slack.go:105]: To differentiate between mud commands, standards commands & admin commands

File: puddle/slack.go (hunk 12 above)
Comment (AlexDunmow) [reply puddle/slack.go:66]: Yeah, probably. Puddle will require a refactor later

File: puddle/slack.go (hunk 11 above)
Comment (nii236) [reply puddle/slack.go:105]: Will there eventually be an authentication system to distinguish between the different roles?

File: puddle/puddle.go (hunk 4 above)
Comment (AlexDunmow) [reply puddle/puddle.go:83]: Only if we want to log a lot of noise, comments left in for brevity

File: .gitignore (hunk 1 above)
Comment (AlexDunmow) [reply .gitignore:26]: It has been checked into git, that's why it was removed

File: puddle/cli.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+		msg := slack.Msg{
+			Text: line,
+		}
+		ProcessMessage(msg)
+		fmt.Print("Puddle> ")
+	}
+}
+
+func PrintCLI(text string) {
+	fmt.Print("Puddle> ")
```
Comment (nii236) [inline puddle/cli.go:25]: This func does print out `text`, should it?

Comment (AlexDunmow) [reply puddle/cli.go:25]: Actually func can be removed...
